python/tskit/trees.py

```diff
@@ -8592,6 +8593,156 @@ def genetic_relatedness_vector(
+            U = Q @ U_hat
+            return U[:,:rank], D[:rank], V[:rank]
+
+        def _genetic_relatedness_vector(
+                ts: tskit.Treesequence,
+                arr: np.ndarray,
+                rows: np.ndarray,
+                cols: np.ndarray,
+                centre: bool = False,
+                windows = None,
+                ) -> np.ndarray:
+            """
+            Wrapper around `tskit.TreeSequence.genetic_relatedness_vector` to support centering in respect to individuals.
+            Multiplies an array to the genetic relatedness matrix of :class:`tskit.TreeSequence`.
+
+            :param tskit.TreeSequence ts: A tree sequence.
+            :param numpy.ndarray arr: The array to multiply. Either a vector or a matrix.
+            :param numpy.ndarray rows: Index of rows of the genetic relatedness matrix to be selected.
+            :param numpy.ndarray cols: Index of cols of the genetic relatedness matrix to be selected. The size should match the row length of `arr`.
+            :param bool centre: Centre the genetic relatedness matrix. Centering happens respect to the `rows` and `cols`. 
+            :param windows: An increasing list of breakpoints between the windows to compute the genetic relatedness matrix in.
+            :return: An array that is the matrix-array product of the genetic relatedness matrix and the array. 
+            :rtype: `np.ndarray`
+            """
+            
+            # maps samples to individuals
+            def sample_individual_sparray(ts: tskit.TreeSequence) -> scipy.sparse.sparray:
+                samples_individual = ts.nodes_individual[ts.samples()]
+                return scipy.sparse.csr_array(
+                        (
+                            np.ones(ts.num_samples),
+                            (np.arange(ts.num_samples), samples_individual)
+                        ),
+                        shape=(ts.num_samples, ts.num_individuals)
+                    )
+            
+            # maps values in idx to num_individuals
+            def individual_idx_sparray(n: int, idx: np.ndarray) -> scipy.sparse.sparray:
+                return scipy.sparse.csr_array(
+                        (
+                            np.ones(idx.size),
+                            (idx, np.arange(idx.size))
+                        ),
+                        shape=(n, idx.size)
+                    )
+                                
+            assert cols.size == arr.shape[0], "Dimension mismatch"
+            # centering
+            x = arr - arr.mean(axis=0) if centre else arr # centering within index in rows
+            x = individual_idx_sparray(ts.num_individuals, cols).dot(x)
+            x = sample_individual_sparray(ts).dot(x)
+            x = ts.genetic_relatedness_vector(W=x, windows=windows, mode="branch", centre=False)
+            x = sample_individual_sparray(ts).T.dot(x)
+            x = individual_idx_sparray(ts.num_individuals, rows).T.dot(x)
```

Comment:
I think this assumes that all individuals' nodes are samples.  Note that we can use the `nodes` argument to `genetic_relatedness_vector` to get an arbitrary list of (possibly non-sample) nodes; why not just use that?

So, I think we can do something like this:
```
ij = np.vstack([[n, k] for k, i in enumerate(individuals) for n in self.individual(i).nodes])
sample_list = ij[:, 0]
indiv_index = ij[:, 1]
x = ts.genetic_relatedness_vector(W=x, ..., nodes=sample_list)
x = np.bincount(indiv_index, x)
```

Reply:
```
x = ts.genetic_relatedness_vector(W=x, ..., nodes=sample_list)
```
should slightly be
```
x = ts.genetic_relatedness_vector(W=x[indiv_index], ..., nodes=sample_list)
```
to expand the array of individuals to array of nodes, I think?

Reply:
hah, yes - good catch!